Proxy Auth support

maweck opened the pull request:
This should more or less solve #471  - As I am a Java guy there may be several things to improve so a review is more than welcome :) 

If this PR is fine I would create some documentation afterwards.

maweck committed:
Change the order of commands to be able to cache more layers in case of multiple builds triggered in a row

maweck committed:
Fix #471

maweck committed:
Format Code

maweck committed:
Revert "Change the order of commands to be able to cache more layers in case of multiple builds triggered in a row"

This reverts commit 01362f34ee45b342f4e9148730ccd30027e5aebf.

eine commented on http/auth.go:
```diff
@@ -51,31 +51,33 @@ func reCaptcha(host, secret, response string) (bool, error) {
-	if err != nil {
-		return http.StatusForbidden, nil
-	}
+	} else if c.AuthMethod == "proxy" {
```
I think this is an unnecessary branch. The previous `if` contains a return statement, so there is no need for an `else` here. The flow will automatically execute `if c.AuthMethod == "proxy"` after evaluating `c.NoAuth` to false. Please, remove `else` and start in a new line.

eine commented on http/auth.go:
```diff
@@ -51,31 +51,33 @@ func reCaptcha(host, secret, response string) (bool, error) {
-	}
+	} else if c.AuthMethod == "proxy" {
+		// Receive the Username from the Header.
+		cred.Username = r.Header.Get(c.LoginHeader)
```
Can you merge the following lines here?

```
	// Checks if the user exists.
	u, err := c.Store.Users.GetByUsername(cred.Username, c.NewFS)
	if err != nil {
		return http.StatusForbidden, nil
    }
```

I.e.:

```
	// Receive the Username from the Header and check if it exists.
	u, err := c.Store.Users.GetByUsername(r.Header.Get(c.LoginHeader), c.NewFS)
	if err != nil {
		return http.StatusForbidden, nil
    }
```

This will allow not to evaluate `c.AuthMethod` again. See below.

Furthermore, if you put the return after the check, you can avoid the last `else`. See above.

```
c.User = u
return printToken(c, w)
```

eine commented on http/auth.go:
```diff
@@ -85,9 +87,11 @@ func authHandler(c *fb.Context, w http.ResponseWriter, r *http.Request) (int, er
-	// Checks if the password is correct.
-	if !fb.CheckPasswordHash(cred.Password, u.Password) {
-		return http.StatusForbidden, nil
+	if c.AuthMethod != "proxy" {
```
If the check and return instructions are added inside the previous block (see comment above), this additional condition needs not to be checked.

eine commented on http/auth.go:
```diff
@@ -51,31 +51,33 @@ func reCaptcha(host, secret, response string) (bool, error) {
-	}
-
-	// Receive the credentials from the request and unmarshal them.
-	var cred cred
```
I think that you removed this line. So trying to use `cred.Username` and/or `cred.ReCaptcha` afterwards should fail.

hacdias commented on cmd/filebrowser/main.go:
```diff
@@ -2,25 +2,22 @@ package main
+	"gopkg.in/natefinch/lumberjack.v2"
+	"io/ioutil"
+	"log"
+	"net"
```
I wonder why all of these imports moved here. 

By the way, @1138-4EB just found out CircleCI had the option to build forked PRs disabled. I just enabled it, but I believe we'll need changes on this PR to trigger it.

eine replied:
They are sorted in alphabetical order now. I think this is because `gofmt` was used. Should we use `goimports` instead? See http://goinbigdata.com/goimports-vs-gofmt/

hacdias reviewed (dismissed):
Seems good overall. Take a look at @1138-4EB's annotations please 😄 And **thanks**!

Could you rebase it on master?

maweck committed:
Adjustment based on the review

hacdias commented on cmd/filebrowser/main.go:
```diff
@@ -63,6 +64,8 @@ func init() {
 	flag.BoolVar(&allowEdit, "allow-edit", true, "Default allow edit option for new users")
 	flag.BoolVar(&allowPublish, "allow-publish", true, "Default allow publish option for new users")
+	flag.StringVar(&auth.method, "auth.method", "default", "Switch between 'default' and 'proxy' authentication.")
+	flag.StringVar(&auth.loginHeader, "auth.login-header", "X-Forwarded-User", "The header name used for proxy authentication.")
```
I don't know how I feel about the dot in the flag honestly. Probably replace it by `-` for consistency. @maweck why did you choose to use `.`?

maweck replied:
Because that is how it is done in most of the tools I work with (Maven, SonarQube, Jenkins, ..... ). In addition it allows you to "group" specific configs. Here is a sample from SonarQube:

>sonar.web.sso.enable=false
>sonar.web.sso.loginHeader=X-Forwarded-Login
>sonar.web.sso.nameHeader=X-Forwarded-Name
>sonar.web.sso.emailHeader=X-Forwarded-Email
>...
>sonar.security.realm=LDAP
>ldap.url=ldap://localhost:10389
>ldap.bindDn=cn=sonar,ou=users,o=mycompany

But sure, I can change that if you want :)

hacdias replied:
Looking at that tools they also seem to use camelCase. It would not be a bad idea. There are some of the other options which should be also grouped.

Could you add them with dots and camelCase please? I'll make another PR to update the other ones because it actually makes sense.

hacdias replied:
Could you also rework it in a way to include noAuth on auth object? Maybe could be renamed auth.disable

hacdias replied:
Ref.: #497

eine reviewed (approved):
The changes I suggested have been applied. It's ok on my side. Just the renaming suggested by @hacdias is pending.

maweck committed:
Rename "login-header" to "loginHeader" and prepare auth.method to accept "none" as a value

maweck committed:
Fixed line break

maweck committed:
Readd "lumberjack.v2" import which was removed by gofmt

Sorry - I do my tests and run "gofmt" before comitting the changes - It sadly seems like it is messing up the imports over and over again.
